File: AkkaBoids/src/main/scala/Boid.scala
package it.unibo.pcd

object Boid:

  case class Position(x: Double, y: Double):
    private def op(other: Position, op: (Position, Position) => Position): Position = op(this, other)
    def +(other: Position): Position = op(other, (a, b) => Position(a.x + b.x, a.y + b.y))
    def -(other: Position): Position = op(other, (a, b) => Position(a.x - b.x, a.y - b.y))
    def *(scalar: Double): Position = Position(x * scalar, y * scalar)
    def /(scalar: Double): Position = Position(x / scalar, y / scalar)
    def distance(other: Position): Double = math.sqrt((other.x - x) * (other.x - x) + (other.y - y) * (other.y - y))
  object Position:
    def zero: Position = Position(0, 0)

  case class Velocity(x: Double, y: Double):
    def +(other: Velocity): Velocity = Velocity(x + other.x, y + other.y)
    def abs: Double = math.sqrt(x * x + y * y)
    def normalized: Velocity = Velocity(x / abs, y / abs)
    def /(scalar: Double): Position = Position(x / scalar, y / scalar)
    def *(scalar: Double): Velocity = Velocity(x * scalar, y * scalar)
  object Velocity:
    def zero: Velocity = Velocity(0, 0)

  case class Boid(position: Position, velocity: Velocity):
    def apply(p: Position, v: Velocity): Boid = Boid(p, v)

    def update(model: BoidsModel): Boid =
      val sep = separation(model) * model.separationWeight
      val ali = alignment(model) * model.alignmentWeight
      val coh = cohesion(model) * model.cohesionWeight

      val newVelocity = velocity + sep + ali + coh
      val speed = newVelocity.abs
      val newVelocityNormalized =
        if speed > model.maxSpeed then newVelocity.normalized * model.maxSpeed else newVelocity

      val newPosition = position + Position(newVelocityNormalized.x, newVelocityNormalized.y)

      Boid(wrapPosition(newPosition, model), newVelocityNormalized)

    private def wrapPosition(pos: Position, model: BoidsModel): Position =
      val newX = pos.x match
        case x if x < model.minX => x + model.width
        case x if x >= model.maxX => x - model.width
        case x => x

      val newY = pos.y match
        case y if y < model.minY => y + model.height
        case y if y >= model.maxY => y - model.height
        case y => y

      Position(newX, newY)

    private def neighbors(implicit model: BoidsModel): List[Boid] =
      for
        boid <- model.boids
        if boid != this && boid.position.distance(position) < model.perceptionRadius
      yield boid

    private def separation(implicit model: BoidsModel): Velocity =
      val boids = for
        boid <- neighbors
        if boid.position.distance(position) < model.avoidRadius
      yield boid.position

      if boids.isEmpty then Velocity.zero
      else
        val res = boids.foldLeft(position)(_ - _) / boids.size
        Velocity(res.x, res.y).normalized

    private def cohesion(implicit mode: BoidsModel): Velocity =
      if neighbors.isEmpty then Velocity.zero
      else
        val center = neighbors.map(_._1).foldLeft(Position.zero)(_ + _) / neighbors.size
        Velocity(center.x - position.x, center.y - position.y).normalized

    private def alignment(implicit mode: BoidsModel): Velocity =
      if neighbors.isEmpty then Velocity.zero
      else
        val avgVel = neighbors.map(_._2).foldLeft(Velocity.zero)(_ + _) / neighbors.size
        Velocity(avgVel.x - velocity.x, avgVel.y - velocity.y).normalized

end Boid

File: AkkaBoids/src/main/scala/legacy/Boid.java
package it.unibo.pcd.legacy;

import it.unibo.pcd.model.BoidsModel;
import it.unibo.pcd.model.Position;
import it.unibo.pcd.model.Velocity;

import java.util.ArrayList;
import java.util.List;

public class Boid {

    private Position pos;
    private Velocity vel;

    public Boid(Position pos, Velocity vel) {
    	this.pos = pos;
    	this.vel = vel;
    }
    
    public Position getPos() {
    	return pos;
    }

    public Velocity getVel() {
    	return vel;
    }
    
    public void update(BoidsModel model, List<Boid> nearbyBoids) {

    	/* change velocity vector according to separation, alignment, cohesion */

    	Velocity separation = calculateSeparation(nearbyBoids, model);
    	Velocity alignment = calculateAlignment(nearbyBoids, model);
    	Velocity cohesion = calculateCohesion(nearbyBoids, model);
    	
    	vel = vel.$plus(alignment.$mul(model.alignmentWeight()))
    			.$plus(separation.$mul(model.separationWeight()))
    			.$plus(cohesion.$mul(model.cohesionWeight()));
        
        /* Limit speed to MAX_SPEED */

        double speed = vel.abs();
        
        if (speed > model.maxSpeed()) {
            vel = vel.normalized().$mul(model.maxSpeed());
        }

        /* Update position */

        pos = pos.$plus(new Position(vel.x(), vel.y()));
        
        /* environment wrap-around */
        
        if (pos.x() < model.minX()) pos = pos.$plus(new Position(model.width(), 0));
        if (pos.x() >= model.maxX()) pos = pos.$plus(new Position(-model.width(), 0));
        if (pos.y() < model.minY()) pos = pos.$plus(new Position(0, model.height()));
        if (pos.y() >= model.maxY()) pos = pos.$plus(new Position(0, -model.height()));
    }

    private Velocity calculateAlignment(List<Boid> nearbyBoids, BoidsModel model) {
        double avgVx = 0;
        double avgVy = 0;
        if (nearbyBoids!= null && !nearbyBoids.isEmpty()) {
	        for (Boid other : nearbyBoids) {
	        	Velocity otherVel = other.getVel();
	            avgVx += otherVel.x();
	            avgVy += otherVel.y();
	        }	        
	        avgVx /= nearbyBoids.size();
	        avgVy /= nearbyBoids.size();
	        return new Velocity(avgVx - vel.x(), avgVy - vel.y()).normalized();
        } else {
        	return new Velocity(0, 0);
        }
    }

    private Velocity calculateCohesion(List<Boid> nearbyBoids, BoidsModel model) {
        double centerX = 0;
        double centerY = 0;
        if (nearbyBoids != null && !nearbyBoids.isEmpty()) {
	        for (Boid other: nearbyBoids) {
	        	Position otherPos = other.getPos();
	            centerX += otherPos.x();
	            centerY += otherPos.y();
	        }
            centerX /= nearbyBoids.size();
            centerY /= nearbyBoids.size();
            return new Velocity(centerX - pos.x(), centerY - pos.y()).normalized();
        } else {
        	return new Velocity(0, 0);
        }
    }
    
    private Velocity calculateSeparation(List<Boid> nearbyBoids, BoidsModel model) {
        double dx = 0;
        double dy = 0;
        int count = 0;
        if (nearbyBoids != null && !nearbyBoids.isEmpty()) {
            for (Boid other: nearbyBoids) {
                Position otherPos = other.getPos();
                double distance = pos.distance(otherPos);
                if (distance < model.avoidRadius()) {
                    dx += pos.x() - otherPos.x();
                    dy += pos.y() - otherPos.y();
                    count++;
                }
            }
        }
        if (count > 0) {
            dx /= count;
            dy /= count;
            return new Velocity(dx, dy).normalized();
        } else {
        	return new Velocity(0, 0);
        }
    }
}
